### PR TITLE
CCE Conformal factor initial data

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds a CCE initial data generator based on choosing an angular conformal factor based on the value of the CCE scalar \beta. This is motivated by an attempt to make the time variable approximately asymptotically inertial.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by moxcodes** [comment]
@keefemitman can you give this branch a try and make sure that it generates similarly good (or better) initial data to what you saw on the branch we were working with for a prototype?
The choice of options that should be most similar to our most recent iteration is probably something like:
```
ConformalFactor:
  AngularCoordinateTolerance: 1e-13
  MaxIterations: 1000
  RequireConvergence: True
  OptimizeL0Mode: True
  UseBetaIntegralEstimate: False
  ConfromalFactorIterationHeuristic: SpinWeight1CoordPerturbation
  UseInputModes: False
  InputModes: []
```

**Comment by keefemitman** [comment]
> @keefemitman can you give this branch a try and make sure that it generates similarly good (or better) initial data to what you saw on the branch we were working with for a prototype?
> The choice of options that should be most similar to our most recent iteration is probably something like:
> 
> ```
> ConformalFactor:
>   AngularCoordinateTolerance: 1e-13
>   MaxIterations: 1000
>   RequireConvergence: True
>   OptimizeL0Mode: True
>   UseBetaIntegralEstimate: False
>   ConfromalFactorIterationHeuristic: SpinWeight1CoordPerturbation
>   UseInputModes: False
>   InputModes: []
> ```

Will do!

File: src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp [hunk 1]
```diff
@@ -185,8 +424,10 @@ struct InitializeJ<false> : public PUP::able {
...
       const Scalar<SpinWeighted<ComplexDataVector, 2>>& boundary_dr_j,
-      const Scalar<SpinWeighted<ComplexDataVector, 0>>& r, size_t l_max,
-      size_t number_of_radial_points) const noexcept = 0;
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& r,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& beta, size_t l_max,
+      size_t number_of_radial_points,
+      gsl::not_null<Parallel::NodeLock*> hdf5_lock) const noexcept = 0;
 };
 }  // namespace InitializeJ
 }  // namespace Cce
```
Comment (keefemitman) [inline src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp:438]: Also need the line `#include "Evolution/Systems/Cce/Initialize/ConformalFactor.hpp"`

File: src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp [hunk 2]
```diff
@@ -163,7 +401,8 @@ template <>
 struct InitializeJ<false> : public PUP::able {
   using boundary_tags = tmpl::list<Tags::BoundaryValue<Tags::BondiJ>,
                                    Tags::BoundaryValue<Tags::Dr<Tags::BondiJ>>,
-                                   Tags::BoundaryValue<Tags::BondiR>>;
+                                   Tags::BoundaryValue<Tags::BondiR>,
+                                   Tags::BoundaryValue<Tags::BondiBeta>>;
 
   using mutate_tags = tmpl::list<Tags::BondiJ, Tags::CauchyCartesianCoords,
                                  Tags::CauchyAngularCoords>;
```
Comment (keefemitman) [inline src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp:410]: Line 413 needs ConformalFactor as a part of the list

**Review by keefemitman — CHANGES_REQUESTED** [review]
I was able to run the code with the new initial data scheme, but I found that I had to lower the angular tolerance to around 1e-4 instead of its suggested value of 1e-11. Is this ok?

**Comment by moxcodes** [comment]
Thanks for checking that @keefemitman -- I would have anticipated that the algorithm would converge better than that. Can you try changing the OptimizeL0 option to `  OptimizeL0Mode: False`?
How does the value that you need to set for the threshold compare to the printed error measure from the prototype branch?

**Comment by keefemitman** [comment]
@moxcodes With `OptimizeL0Mode` as false I still fail to obtain the specified tolerance. The print out is
"final maximum over collocation points deviation of J from target of 1.26732e-05"

**Comment by moxcodes** [comment]
The high tolerance is concerning, but I think not necessarily a deal-breaker if the initial data still seems to be an improvement in terms of the transients we were seeing. Do you have any sense for how the waveform on this one compares with the waveforms you were seeing from the `cce_beta_based_id` branch?

**Comment by keefemitman** [comment]
I'm still running a waveform right now, but I'll let you know once it finishes!

**Comment by keefemitman** [comment]
Ok so it looks like the previous version of the conformal factor branch is much better than the new one in that the previous version of the branch largely reduced the transient effects at early times, but the newer version doesn't
Screen Shot 2021-08-23 at 4 11 10 PM

**Comment by moxcodes** [comment]
Thanks, @keefemitman , I'll look into that and update this PR once I manage to find and fix the bug I must have introduced in the new version.

**Comment by moxcodes** [comment]
I pushed a fixup and rebased. I believe that I've fixed the regression -- @keefemitman could you please test again and see if the behavior is now similar to the previous branch that has better initial transient results?
Thanks!

**Comment by keefemitman** [comment]
@moxcodes Unfortunately I'm still seeing the same behavior as before, even with the new changes you made.

**Comment by keefemitman** [comment]
> @moxcodes Unfortunately I'm still seeing the same behavior as before, even with the new changes you made.

Scratch that, this was just due to a different tolerance that I set in the input file. Everything looks good now!

**Comment by kidder** [comment]
@nilsdeppe this has been rebased and sqaushed

File: src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        "m-varies-fastest order."};
+  };
+
+  using options =
+      tmpl::list<AngularCoordinateTolerance, MaxIterations, RequireConvergence,
+                 OptimizeL0Mode, UseBetaIntegralEstimate,
+                 ConformalFactorIterationHeuristic, UseInputModes,
+                 Options::Alternatives<tmpl::list<InputModesFromFile>,
+                                       tmpl::list<InputModes>>>;
+  static constexpr Options::String help = {""};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp]: Maybe have some sort of non-empty string?

File: src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+  WRAPPED_PUPable_decl_template(ConformalFactor);  // NOLINT
+  explicit ConformalFactor(CkMigrateMessage* /*unused*/) {}
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp]: move definition to cpp and do `msg) : InitializeJ<false>(msg) {}`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM, couple easy suggestions, rebase and squash right away please

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**